homeassistant/components/ecovacs_mqtt/config_flow.py

```diff
@@ -0,0 +1,271 @@
+
+from .const import CONF_COUNTRY, DOMAIN, SELF_HOSTED_CONFIGURATION, Mode
+from .util import get_client_device_id
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EcovacsConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Ecovacs."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        self._data: dict[str, Any] = {}
+        self._devices: list[DeviceInfo] = []
+        self._entry: ConfigEntry | None = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> EcovacsOptionsFlowHandler:
+        """Get the options flow for this handler."""
+        return EcovacsOptionsFlowHandler(config_entry)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if not self.show_advanced_options:
+            self._data[CONF_MODE] = Mode.CLOUD
+            return await self.async_step_auth()
+
+        if user_input:
+            self._data.update(user_input)
+            auth_input = None
+            if user_input[CONF_MODE] == Mode.SELF_HOSTED:
+                auth_input = SELF_HOSTED_CONFIGURATION.copy()
+
+            return await self.async_step_auth(user_input=auth_input)
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_MODE, default=Mode.CLOUD
+                    ): selector.SelectSelector(
+                        selector.SelectSelectorConfig(
+                            translation_key="mode", options=list(Mode)
+                        )
+                    )
+                }
+            ),
+        )
+
+    async def async_step_auth(
+        self, user_input: Mapping[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the auth step."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                self._devices = await _retrieve_devices(self.hass, user_input)
+            except ClientError:
+                _LOGGER.debug("Cannot connect", exc_info=True)
+                errors["base"] = "cannot_connect"
+            except InvalidAuthenticationError:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if not errors:
+                self._data.update(user_input)
+                if self._entry:
+                    # reauthentication
+                    self.hass.config_entries.async_update_entry(
+                        self._entry, data=self._data
+                    )
+                    self.hass.async_create_task(
+                        self.hass.config_entries.async_reload(self._entry.entry_id)
+                    )
```

Comment:
You already reload on changing data, so you don't need this. (Or it will reload twice iirc)